File: OpenDrop.Infrastructure/Transfer/TransferServer.cs
using System.Collections.Concurrent;
using System.Net;
using System.Security.Cryptography;
using OpenDrop.Core.Crypto;
using OpenDrop.Core.Security;
using OpenDrop.Core.Transfer;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace OpenDrop.Infrastructure.Transfer;

public sealed class TransferServer : IAsyncDisposable
{
    private readonly ILogger<TransferServer> _log;
    private readonly TransferServerOptions _options;
    private readonly IKex _kex;
    private readonly IKeyDerivation _kdf;
    private readonly IAead _aead;

    private readonly ConcurrentDictionary<string, TransferSession> _sessions = new(StringComparer.Ordinal);
    private WebApplication? _app;
    private Task? _runTask;

    // Basic abuse protection: per-remote-IP request tokens.
    private readonly TokenBucketLimiter _limiter = new(capacity: 60, refillPerSecond: 30); // burst 60, ~30 req/s sustained

    public event EventHandler<TransferOffer>? OfferReceived;

    public TransferServer(
        TransferServerOptions options,
        IKex kex,
        IKeyDerivation kdf,
        IAead aead,
        ILogger<TransferServer> log)
    {
        _options = options;
        _kex = kex;
        _kdf = kdf;
        _aead = aead;
        _log = log;
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        if (_app is not null) return;

        Directory.CreateDirectory(_options.StorageDirectory);

        var builder = WebApplication.CreateBuilder(new WebApplicationOptions
        {
            ApplicationName = typeof(TransferServer).Assembly.FullName,
            Args = Array.Empty<string>(),
            EnvironmentName = Environments.Production
        });

        builder.Logging.ClearProviders(); // app should configure logging; keep minimal here

        builder.WebHost.ConfigureKestrel(k =>
        {
            k.Listen(IPAddress.Parse(_options.BindAddress), _options.Port);
        });

        var app = builder.Build();

        app.Use(async (ctx, next) =>
        {
            var key = ctx.Connection.RemoteIpAddress?.ToString() ?? "unknown";
            if (!_limiter.TryConsume(key, 1))
            {
                ctx.Response.StatusCode = StatusCodes.Status429TooManyRequests;
                return;
            }
            await next();
        });

        app.MapPost("/api/v1/sessions", async (HttpContext ctx) =>
        {
            var req = await ctx.Request.ReadFromJsonAsync<CreateSessionRequest>(cancellationToken: ctx.RequestAborted);
            if (req is null) return Results.BadRequest();

            if (req.Offer.FileSize <= 0 || req.Offer.FileSize > _options.MaxFileSizeBytes)
                return Results.BadRequest("Invalid file size.");

            var ext = Path.GetExtension(req.Offer.FileName);
            if (_options.AllowedExtensions.Length > 0 &&
                !string.IsNullOrWhiteSpace(ext) &&
                !_options.AllowedExtensions.Contains(ext, StringComparer.OrdinalIgnoreCase))
            {
                return Results.BadRequest("File type not allowed.");
            }

            // Generate receiver ephemeral key for this session.
            var recvKp = _kex.CreateIdentityKeyPair();
            var salt = RandomNumberGenerator.GetBytes(16);
            var (key, baseNonce) = TransferCrypto.DeriveAeadKeyAndNonce(
                _kex, _kdf,
                myPriv: recvKp.PrivateKey,
                peerPub: req.SenderEphemeralPublicKey,
                salt: salt,
                infoLabel: "airdroplike/v1/session");

            var sessionId = Guid.NewGuid().ToString("n");
            var safeName = PathSafety.SanitizeFileName(req.Offer.FileName);
            var tempPath = Path.Combine(_options.StorageDirectory, $"{sessionId}.{safeName}.part");

            var offer = req.Offer with
            {
                SessionId = sessionId,
                CreatedUtc = DateTimeOffset.UtcNow,
                FileName = safeName
            };

            var session = new TransferSession
            {
                Offer = offer,
                SessionId = sessionId,
                SenderEphemeralPublicKey = req.SenderEphemeralPublicKey,
                ReceiverEphemeralPrivateKey = recvKp.PrivateKey,
                ReceiverEphemeralPublicKey = recvKp.PublicKey,
                Salt = salt,
                AeadKey = key,
                BaseNonce = baseNonce,
                TempPath = tempPath,
                BytesReceived = 0,
                Finalized = false
            };

            _sessions[sessionId] = session;

            OfferReceived?.Invoke(this, offer);
            _log.LogInformation("Offer received: {Name} ({Bytes} bytes) session {SessionId}.", offer.FileName, offer.FileSize, sessionId);

            return Results.Ok(new CreateSessionResponse(sessionId, recvKp.PublicKey, salt));
        });

        app.MapGet("/api/v1/sessions/{sessionId}/status", (string sessionId) =>
        {
            if (!_sessions.TryGetValue(sessionId, out var s)) return Results.NotFound();
            return Results.Ok(new TransferResumeState(sessionId, s.BytesReceived, s.Finalized));
        });

        app.MapPut("/api/v1/sessions/{sessionId}/chunks", async (HttpContext ctx, string sessionId) =>
        {
            if (!_sessions.TryGetValue(sessionId, out var s)) return Results.NotFound();
            if (s.Finalized) return Results.BadRequest("Already finalized.");

            if (!long.TryParse(ctx.Request.Query["offset"], out var offset) || offset < 0)
                return Results.BadRequest("Missing/invalid offset.");
            if (!long.TryParse(ctx.Request.Query["index"], out var index) || index < 0)
                return Results.BadRequest("Missing/invalid index.");

            var declaredSha = ctx.Request.Query["sha256"].ToString();
            if (string.IsNullOrWhiteSpace(declaredSha)) return Results.BadRequest("Missing sha256.");

            if (offset != s.BytesReceived)
                return Results.Conflict("Offset mismatch.");

            using var ms = new MemoryStream();
            await ctx.Request.Body.CopyToAsync(ms, ctx.RequestAborted);
            var enc = ms.ToArray();

            if (!string.Equals(TransferCrypto.Sha256Hex(enc), declaredSha, StringComparison.OrdinalIgnoreCase))
                return Results.BadRequest("Chunk hash mismatch.");

            var nonce = TransferCrypto.NonceForChunk(s.BaseNonce, index);
            byte[] pt;
            try
            {
                // AAD binds the chunk location to the ciphertext.
                var aad = System.Text.Encoding.UTF8.GetBytes($"{sessionId}:{offset}:{index}");
                pt = _aead.Decrypt(s.AeadKey, nonce, enc, aad);
            }
            catch (CryptographicException)
            {
                return Results.Unauthorized();
            }

            if (s.BytesReceived + pt.Length > s.Offer.FileSize)
                return Results.BadRequest("Too much data.");

            Directory.CreateDirectory(Path.GetDirectoryName(s.TempPath)!);
            await using (var fs = new FileStream(s.TempPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read))
            {
                fs.Seek(offset, SeekOrigin.Begin);
                await fs.WriteAsync(pt, ctx.RequestAborted);
                await fs.FlushAsync(ctx.RequestAborted);
            }

            s.BytesReceived += pt.Length;
            return Results.NoContent();
        });

        app.MapPost("/api/v1/sessions/{sessionId}/finalize", async (string sessionId) =>
        {
            if (!_sessions.TryGetValue(sessionId, out var s)) return Results.NotFound();
            if (s.Finalized) return Results.NoContent();

            if (s.BytesReceived != s.Offer.FileSize)
                return Results.Conflict("Incomplete.");

            // Verify final hash.
            await using var fs = new FileStream(s.TempPath, FileMode.Open, FileAccess.Read, FileShare.Read);
            var hash = await SHA256.HashDataAsync(fs, cancellationToken);
            var hex = Convert.ToHexString(hash).ToLowerInvariant();
            if (!string.Equals(hex, s.Offer.Sha256Hex, StringComparison.OrdinalIgnoreCase))
                return Results.BadRequest("File hash mismatch.");

            var finalPath = Path.Combine(_options.StorageDirectory, s.Offer.FileName);
            finalPath = EnsureUnique(finalPath);
            File.Move(s.TempPath, finalPath);
            s.Finalized = true;

            _log.LogInformation("Transfer finalized: {Path}", finalPath);
            return Results.Ok(new { path = finalPath });
        });

        _app = app;
        _runTask = app.RunAsync(cancellationToken);
        await Task.CompletedTask;
    }

    private static string EnsureUnique(string path)
    {
        if (!File.Exists(path)) return path;
        var dir = Path.GetDirectoryName(path)!;
        var name = Path.GetFileNameWithoutExtension(path);
        var ext = Path.GetExtension(path);
        for (int i = 1; i < 10_000; i++)
        {
            var candidate = Path.Combine(dir, $"{name} ({i}){ext}");
            if (!File.Exists(candidate)) return candidate;
        }
        throw new IOException("Unable to pick unique filename.");
    }

    public async ValueTask DisposeAsync()
    {
        if (_app is null) return;
        try
        {
            await _app.StopAsync();
        }
        catch { /* ignore */ }
        await _app.DisposeAsync();
        _app = null;
        if (_runTask is not null) await _runTask;
    }

    private sealed record CreateSessionRequest(TransferOffer Offer, byte[] SenderEphemeralPublicKey);
    private sealed record CreateSessionResponse(string SessionId, byte[] ReceiverEphemeralPublicKey, byte[] Salt);
}

File: pyqt6_opendrop/opendrop_pyqt6/transfer_server.py
from __future__ import annotations

import hashlib
import os
import threading
import uuid
from dataclasses import dataclass
from pathlib import Path

from flask import Flask, jsonify, request
from werkzeug.serving import make_server

from .crypto import compute_shared_secret, create_keypair, decrypt_chunk, derive_key_and_base_nonce, nonce_for_chunk, sha256_hex
from .security import sanitize_file_name


@dataclass(slots=True)
class TransferSession:
    session_id: str
    file_name: str
    file_size: int
    sha256_hex_expected: str
    temp_path: Path
    key: bytes
    base_nonce: bytes
    bytes_received: int = 0
    finalized: bool = False


class TransferServer:
    def __init__(self, host: str, port: int, storage_dir: Path, max_file_size: int = 4 * 1024 * 1024 * 1024) -> None:
        self.host = host
        self.port = port
        self.storage_dir = storage_dir
        self.max_file_size = max_file_size
        self.sessions: dict[str, TransferSession] = {}
        self.app = Flask("opendrop-transfer")
        self._server = None
        self._thread: threading.Thread | None = None
        self._register_routes()

    def _register_routes(self) -> None:
        @self.app.post("/api/v1/sessions")
        def create_session():
            req = request.get_json(force=True)
            offer = req["offer"]
            if offer["file_size"] <= 0 or offer["file_size"] > self.max_file_size:
                return jsonify({"error": "invalid size"}), 400

            recv_keys = create_keypair()
            salt = os.urandom(16)
            shared = compute_shared_secret(recv_keys.private_key, bytes.fromhex(req["sender_ephemeral_public_key_hex"]))
            key, nonce = derive_key_and_base_nonce(shared, salt)

            session_id = uuid.uuid4().hex
            safe_name = sanitize_file_name(offer["file_name"])
            temp_path = self.storage_dir / f"{session_id}.{safe_name}.part"
            self.sessions[session_id] = TransferSession(
                session_id=session_id,
                file_name=safe_name,
                file_size=int(offer["file_size"]),
                sha256_hex_expected=offer["sha256_hex"],
                temp_path=temp_path,
                key=key,
                base_nonce=nonce,
            )
            return jsonify(
                {
                    "session_id": session_id,
                    "receiver_ephemeral_public_key_hex": recv_keys.public_bytes.hex(),
                    "salt_hex": salt.hex(),
                }
            )

        @self.app.get("/api/v1/sessions/<session_id>/status")
        def status(session_id: str):
            s = self.sessions.get(session_id)
            if not s:
                return jsonify({"error": "not found"}), 404
            return jsonify({"session_id": s.session_id, "bytes_received": s.bytes_received, "finalized": s.finalized})

        @self.app.put("/api/v1/sessions/<session_id>/chunks")
        def chunk(session_id: str):
            s = self.sessions.get(session_id)
            if not s:
                return jsonify({"error": "not found"}), 404
            offset = int(request.args.get("offset", "-1"))
            index = int(request.args.get("index", "-1"))
            declared_sha = request.args.get("sha256", "")
            if offset != s.bytes_received:
                return jsonify({"error": "offset mismatch"}), 409
            enc = request.get_data()
            if sha256_hex(enc) != declared_sha:
                return jsonify({"error": "chunk hash mismatch"}), 400
            aad = f"{session_id}:{offset}:{index}".encode("utf-8")
            plain = decrypt_chunk(s.key, nonce_for_chunk(s.base_nonce, index), enc, aad)
            if s.bytes_received + len(plain) > s.file_size:
                return jsonify({"error": "too much data"}), 400
            s.temp_path.parent.mkdir(parents=True, exist_ok=True)
            with s.temp_path.open("ab") as f:
                f.write(plain)
            s.bytes_received += len(plain)
            return ("", 204)

        @self.app.post("/api/v1/sessions/<session_id>/finalize")
        def finalize(session_id: str):
            s = self.sessions.get(session_id)
            if not s:
                return jsonify({"error": "not found"}), 404
            if s.bytes_received != s.file_size:
                return jsonify({"error": "incomplete"}), 409
            digest = hashlib.sha256(s.temp_path.read_bytes()).hexdigest()
            if digest != s.sha256_hex_expected:
                return jsonify({"error": "file hash mismatch"}), 400
            final_path = self.storage_dir / s.file_name
            i = 1
            while final_path.exists():
                final_path = self.storage_dir / f"{Path(s.file_name).stem} ({i}){Path(s.file_name).suffix}"
                i += 1
            s.temp_path.replace(final_path)
            s.finalized = True
            return jsonify({"path": str(final_path)})

    def start(self) -> None:
        self.storage_dir.mkdir(parents=True, exist_ok=True)
        self._server = make_server(self.host, self.port, self.app)
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        if self._server:
            self._server.shutdown()
